src/lseq/ident.rs

```diff
@@ -0,0 +1,409 @@
+/// two sites decide to pick the same child index to allocate a fresh node
+#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Clone, Serialize, Deserialize, Hash)]
+pub struct Identifier<A: Actor> {
+    path: Vec<(u64, Option<A>)>,
+    // site_id: u64,
+    // counter: u64
+}
+
+impl<A: Actor> Identifier<A> {
+    /// Get the size of an identifier.
+    pub fn len(&self) -> usize {
+        self.path.len()
+    }
+
+    /// Check if the identifier is empty.
+    pub fn is_empty(&self) -> bool {
+        self.path.is_empty()
+    }
+}
+
+/// A generator for fresh identifiers.
+///
+/// These identifiers represent a path in an exponential tree. At each level of the tree the amount
+/// of children doubles. This means that we can store a lot of nodes in a very low height!
+///
+/// The [`crate::LSeq`] tree also has an additional restriction: at each level the minimum and maximum nodes
+/// cannot be chosen when allocating fresh nodes. This is to ensure there is always a free node
+/// that can be used to create a lower level.
+//#[derive(Serialize, Deserialize)]
+pub struct IdentGen<A: Actor> {
+    initial_base_bits: u32,
+    strategy_vec: BitVec,
+    /// Site id of this tree
+    pub site_id: A,
+    // clock: u64
+}
+
+impl<A: Actor> IdentGen<A> {
+    /// Create a fresh tree with 0 node.
+    pub fn new(site_id: A) -> Self {
+        Self::new_with_args(INITIAL_BASE, site_id)
+    }
+
+    /// Create a tree with a custom initial size.
+    pub fn new_with_args(base: u32, site_id: A) -> Self {
+        IdentGen {
+            initial_base_bits: base,
+            strategy_vec: BitVec::new(),
+            site_id,
+        }
+    }
+
+    /// The smallest possible node in a tree.
+    pub fn lower(&self) -> Identifier<A> {
+        Identifier {
+            path: vec![(0, None)],
+        }
+    }
+
+    /// The absolute largest possible node.
+    pub fn upper(&self) -> Identifier<A> {
+        Identifier {
+            path: vec![(2u64.pow(self.initial_base_bits) - 1, None)],
+        }
+    }
+
+    /// Allocates a new identifier between `p` and `q` such that `p < z < q`.
+    ///
+    /// The way to think about this is to consider the problem of finding a short number between
+    /// two decimal numbers. For example, between `0.2` and `0.4` we'd like to return `0.3`.
+    /// But what about `0.2` and `0.3`? Well a nice property is that if we add a digit to `0.2`, the
+    /// number produced will be larger than `0.2` but smaller than `0.3`.
+    ///
+    /// If you view identifiers `p` and `q` as decimal numbers `0.p` and `0.q` then all we're doing is
+    /// finding a number between them!
+    ///
+    /// # Panics
+    ///
+    /// * `p` equal to `q`.
+    /// * If no identifier of length less than 31 exists between `p` and `q`.
+    pub fn alloc(&mut self, p: &Identifier<A>, q: &Identifier<A>) -> Identifier<A> {
+        assert!(p != q, "allocation range should be non-empty");
+        if q < p {
+            self.alloc(q, p);
```

Comment:
opa, little bug here, we should be returning otherwise we fall through the if and continue allocating with Q < P
```suggestion
            return self.alloc(q, p);
```

I was reworking the property tests for this module and found this one, I'm still trying to find how we were missing this with the tests you had in place since it seems we should have been catching it :thinking:.

Reply:
Superseded by https://github.com/xldenis/rust-crdt/pull/6